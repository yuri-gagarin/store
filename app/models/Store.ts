import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;