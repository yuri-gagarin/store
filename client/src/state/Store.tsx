import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

