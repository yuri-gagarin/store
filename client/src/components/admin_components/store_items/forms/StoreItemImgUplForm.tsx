import React, { useState, useContext } from "react";
import { Button, Form } from "semantic-ui-react";
// css imports //
import "./css/storeItemImgUploadForm.css";
// actions //
import { uploadStoreItemImage } from "../actions/APIStoreItemActions";
// state //
import { Store } from "../../../../state/Store";

const StoreItemImageUplForm: React.FC<{}> = (props): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentStoreItemData } = state.storeItemState;
  // local component state //
  const [ file, setFile ] = useState<File>();
  const [ imgUploadError, setImgUploadError ] = useState<boolean>(false);
  // event handlers and listeners //
  const handleButtonClick = () => {

  };
  const uploadFile = () => {
    if (file) {
      const formData = new FormData();
      const { _id } = currentStoreItemData;
      formData.append("storeItemImage", file);
      uploadStoreItemImage(_id, formData, state, dispatch)
        .then((_) => {
          setFile(undefined);
          setImgUploadError(false);
        })
        .catch((_) => {
          // handle error ? show error screen? for later ... //
          setImgUploadError(true);
        });
    }
  };
  const cancelFile = () => {
    setFile(undefined);
  };
  const fileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setFile(e.target.files[0])
    }
  };

  return (
    <div id="storeItemImgUplFormHolder">
      <div><p>Image Uploader</p></div>
      <Form id="storeItemImgUploadForm">
        {
          file ?  
          <div id="storeItemImgUploadControlls">
            {
              !imgUploadError ?
              <Button 
                id="storeItemImgUploadBtn"
                as="label"
                content="Upload"
                icon="upload"
                onClick={uploadFile}
                loading={loading}
              />
              :
              <Button 
                id="storeItemImgUplRetryBtn"
                as="label"
                content="Cancel"
                labelPosition="left"
                icon="cancel"
                onClick={cancelFile}
                loading={loading}
              />
            }
            <Button 
              id="storeItemImgCancelBtn"
              as="label"
              content="Cancel"
              labelPosition="left"
              icon="cancel"
              onClick={cancelFile}
            />
          </div>
          : 
          <div id="storeItemImgInputControlls"> 
            <Button
              id="storeItemImgSelectBtn"
              as="label"
              content="Choose Image"
              labelPosition="left"
              htmlFor="storeItemImgFile"
              icon="file"
              onClick={handleButtonClick}
            />
            <input type="file" id="storeItemImgFile" hidden onChange={fileChange} />
          </div>
        }
       

       <span>{file ? file.name : "No File"}</span>
      </Form>
    </div>
  );
};

export default StoreItemImageUplForm;

