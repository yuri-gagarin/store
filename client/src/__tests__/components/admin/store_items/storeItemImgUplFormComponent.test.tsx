import React from "react";
import { Button } from "semantic-ui-react";
import moxios from "moxios";
// testing imports //
import { mount, shallow, ReactWrapper, ShallowWrapper } from "enzyme";
import { act } from 'react-dom/test-utils';
// components //
import StoreItemImgUplForm from "../../../../components/admin_components/store_items/forms/StoreItemImgUplForm";
// React.Context and State //
import { TestStateProvider } from "../../../../state/Store";
// helpers //
import MockFile from "../../../helpers/mockFile";
import { createMockStoreItemImage, createMockStoreItems } from "../../../../test_helpers/storeItemHelpers";
import { generateCleanState } from "../../../../test_helpers/miscHelpers";

describe("StoreItem Image Upload Form Tests", () => {
  let mockStoreItemData: IStoreItemData;

  beforeAll(() => {
    mockStoreItemData = createMockStoreItems(1)[0];
    mockStoreItemData.images[0] = createMockStoreItemImage(mockStoreItemData._id)
  });

  describe("Render tests without any Image data", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      wrapper = mount(<StoreItemImgUplForm />)
     });

    it("Should properly render", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render one '#storeItemImgSelectBtn' element", () => {
      const selectImgBtn = wrapper.render().find("#storeItemImgSelectBtn");
      expect(selectImgBtn.length).toEqual(1);
    });
    it("Should NOT render '#storeItemImgUplControls' element", () => {
      const controls = wrapper.render().find("#storeItemImgUplControls");
      expect(controls.length).toEqual(0);
    });
    it("Should have an input", () => {
      const input = wrapper.find("input");
      expect(input.length).toEqual(1);
    });
  });
  
  describe("Render tests with an Image file present", () => {
    let wrapper: ReactWrapper;
    let input: ReactWrapper;

    beforeAll(() => {
      wrapper = mount(<StoreItemImgUplForm />);
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = wrapper.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    it("Should properly set the file in props", () => {
      expect(input.prop("type")).toBeDefined();
      expect(input.prop("type")).toEqual("file");
    });
    it("Should successfully update and render the '#storeItemImgUploadlBtn", () => {
      const uplBtn = wrapper.render().find("#storeItemImgUploadBtn");
      expect(uplBtn.length).toEqual(1);
    });
    it("Should successfully update and render the '#storeItemImgCancelBtn'", () => {
      const cancelBtn = wrapper.render().find("#storeItemImgCancelBtn");
      expect(cancelBtn.length).toEqual(1);
    });
    it("Should NOT render the '#storeItemImgRetryUplBtn'", () => {
      const retryBtn = wrapper.render().find("#storeItemImgRetryUplBtn");
      expect(retryBtn.length).toEqual(0)
    });
    it("Should properly handle the '#storeItemImgCancelBtn' click and render correct local state", () => {
      const imgCancelBtn = wrapper.find("#storeItemImgCancelBtn");
      imgCancelBtn.at(0).simulate("click");
      // assert correct rendering //
      expect(wrapper.render().find("#storeItemImgCancelBtn").length).toEqual(0);
      expect(wrapper.render().find("#storeItemImgUploadBtn").length).toEqual(0);
      expect(wrapper.render().find("#storeItemImgSelectBtn").length).toEqual(1);
    });
  });
  // MOCK Successful Image upload tests //
  describe("'#storeItemImgUploadBtn' functionality and successful upload and local state changes", () => {
    let wrapper: ReactWrapper;
    let input: ReactWrapper;

    beforeAll(() => {
      const state = generateCleanState();
      state.storeItemState.currentStoreItemData = mockStoreItemData;

      moxios.install();
      moxios.stubRequest(`/api/uploads/store_item_images/${mockStoreItemData._id}`, {
        status: 200,
        response: {
          responseMsg: "ok",
          updatedStoreItem: mockStoreItemData
        }
      });

      wrapper = mount(
        <TestStateProvider mockState={state}>
          <StoreItemImgUplForm />
        </TestStateProvider>
      );
      // mock a file attached //
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = wrapper.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    afterEach(() => {
      moxios.uninstall();
    });
    it("Should dispatch a 'correct' API request show 'loader'", async () => {
      const promise = Promise.resolve();
      const imgUpoadBtn = wrapper.find("#storeItemImgUploadBtn");
      imgUpoadBtn.at(0).simulate("click");
      // clear promises //
      await act( async () => promise);
      const uploadButton = wrapper.find(StoreItemImgUplForm).find(Button).at(0);
      expect(uploadButton.props().loading).toEqual(true);
    });
    it("Should correctly render '#storeItemImgSelectBtn' button after 'successful' upload", () => {
      wrapper.update();
      const selectImgBtn = wrapper.find(StoreItemImgUplForm).render().find("#storeItemImgSelectBtn");
      //console.log(wrapper.debug())
      expect(selectImgBtn.length).toEqual(1);
    });
    it("Should NOT render '#storeItemImgRetryBtn' button after 'successful' upload", () => {
      const imgUpoadBtn = wrapper.find(StoreItemImgUplForm).render().find("#storeItemImgRetryBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    });
    it("Should NOT render '#storeItemImgUploadBtn' button after 'successful' upload", () => {
      const imgUpoadBtn = wrapper.find(StoreItemImgUplForm).render().find("#storeItemImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    });
  });
  // END Mock successful Image upload tests //
  // MOCK unsuccessful Image upload tests //
  describe("'#storeItemImgCancelBtn' functionality and a failed upload", () => {
    let wrapper: ReactWrapper;
    let input: ReactWrapper;
    const error = new Error("An Error Occured");

    beforeAll(() => {
      const state = generateCleanState();
      state.storeItemState.currentStoreItemData = mockStoreItemData;

      moxios.install();
      moxios.stubRequest(`/api/uploads/store_item_images/${mockStoreItemData._id}`, {
        status: 500,
        response: {
          responseMsg: "Error",
          error: error
        }
      });

      wrapper = mount(
        <TestStateProvider mockState={state}>
          <StoreItemImgUplForm />
        </TestStateProvider>
      );
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = wrapper.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    
    afterEach(() => {
      moxios.uninstall();
    });

    it("Should successfully handle the '#storeItemImgUploadBtn' API error, render 'loader'", async () => {
      const promise = Promise.resolve();
      const imgUpoadBtn = wrapper.find("#storeItemImgUploadBtn").at(0);

      imgUpoadBtn.simulate("click");
      // loader should be displayed on the  #storeItemImgUploadBtn //
      await act( async() => promise);
      const updatedButton = wrapper.find(StoreItemImgUplForm).find(Button).at(0);
      expect(updatedButton.props().loading).toEqual(true);
    });
    it("Should render '#storeItemImgRetryButton' button after a 'failed' upload", () => {
      wrapper.update();
      const imgUpoadBtn = wrapper.find(StoreItemImgUplForm).render().find("#storeItemImgRetryUplBtn");
      expect(imgUpoadBtn.length).toEqual(1);
    });
    it("Should render '#storeItemImgCancelBtn' button after a 'failed' upload", () => {
      const imgUpoadBtn = wrapper.find(StoreItemImgUplForm).render().find("#storeItemImgCancelBtn");
      expect(imgUpoadBtn.length).toEqual(1);
    });
    it("Should NOT render '#storeItemImgSelectBtn' button after a 'failed' upload", () => {
      const selectImgBtn = wrapper.find(StoreItemImgUplForm).render().find("#storeItemImgSelectBtn");
      expect(selectImgBtn.length).toEqual(0);
    });
  });
  // END Mock unsuccessful Image upload tests //
});