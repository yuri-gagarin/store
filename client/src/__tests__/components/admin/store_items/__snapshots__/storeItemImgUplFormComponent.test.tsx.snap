// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreItem Image Upload Form Tests Render tests without any Image data Should properly render 1`] = `
<div
  id="storeItemImgUplFormHolder"
>
  <div>
    <p>
      Image Uploader
    </p>
  </div>
  <Form
    as="form"
    id="storeItemImgUploadForm"
  >
    <div
      id="storeItemImgInputControlls"
    >
      <Button
        as="label"
        content="Choose Image"
        htmlFor="storeItemImgFile"
        icon="file"
        id="selectStoreItemImgBtn"
        labelPosition="left"
        onClick={[Function]}
      />
      <input
        hidden={true}
        id="storeItemImgFile"
        onChange={[Function]}
        type="file"
      />
    </div>
    <span>
      No File
    </span>
  </Form>
</div>
`;
